pkg/kubelet/status/status_manager.go

```diff
@@ -619,6 +621,16 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 		return
 	}
 
+	// Delay transitioning a pod to a terminal status unless the pod is actually
+	// complete. The Kubelet should never transition a pod to terminal status that could
+	// have running containers and thus actively be leveraging exclusive resources.
+	if !isPhaseTerminal(pod.Status.Phase) && isPhaseTerminal(status.status.Phase) {
+		if m.podDeletionSafety.PodCouldHaveRunningContainers(pod) {
```

Comment:
Just to understand why are we only using `PodCouldHaveRunningContainers` which only checks `CouldHaveRunningContainers` instead of [`PodResourcesAreReclaimed`](https://github.com/kubernetes/kubernetes/blob/043611723136d30ec76782f96c5bb03fe7a47ca4/pkg/kubelet/kubelet_pods.go#L920)  which is a stronger check that also validates that all containers are exited, no pod volumes exist, cgroup is deleted etc?

Reply:
PodResourcesAreReclaimed is “fully terminated”, ie syncTerminatedPod has completed.  The issue in this case is terminating containers, which corresponds with exiting syncTerminatingPods.  As mentioned in slack, if you start a new pod with volumes while volumes used by a previous pod are cleaning up, we gracefully wait already.  So there’s no impact to an end user by accepting that pod early.  We actually need that behavior for certain types of zero disruption upgrades of things that are per node.

The reclaimed check is to cover deletion of a pod, where we need to ensure another node can safely and exclusively access some volume types - unmount ensures no more data can be written to that disk (well, it gives the volume plugin the chance to provide that semantic).  If we didn’t wait, we wouldn’t guarantee the pod safety rule that a volume the scheduler believes is uniquely assigned to a specific node can’t accept writes from another node.

Reply:
Added an extended comment to this effect in the updated PR.  Let me know if that's sufficiently clear (I plan on trying to boil this into a KEP that clarifies the current state and proposes the improvements that could be added, like additional conditions).